gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/PasswordPolicyRequestParseHandler.java

```diff
@@ -43,31 +47,35 @@ public class PasswordPolicyRequestParseHandler extends UserRequestHandler {
     private final PasswordService passwordService;
     private final PasswordPolicyManager passwordPolicyManager;
     private final IdentityProviderManager identityProviderManager;
+    private final AuditService auditService;
 
-    public PasswordPolicyRequestParseHandler(PasswordService passwordService, PasswordPolicyManager passwordPolicyManager, IdentityProviderManager identityProviderManager, Domain domain) {
+    public PasswordPolicyRequestParseHandler(PasswordService passwordService, PasswordPolicyManager passwordPolicyManager, IdentityProviderManager identityProviderManager, Domain domain, AuditService auditService) {
         this.identityProviderManager = identityProviderManager;
         this.passwordPolicyManager = passwordPolicyManager;
         this.passwordService = passwordService;
         this.domain = domain;
+        this.auditService = auditService;
     }
 
     @Override
     public void handle(RoutingContext context) {
         HttpServerRequest request = context.request();
         String password = request.getParam(ConstantKeys.PASSWORD_PARAM_KEY);
         MultiMap queryParams = RequestUtils.getCleanedQueryParams(request);
-
+        Client client = context.get(ConstantKeys.CLIENT_CONTEXT_KEY);
+        User user = getUser(context, client);
+        String source = Optional.ofNullable(user.getSource())
+                .orElseGet(() -> UserRegistrationIdpResolver.getRegistrationIdp(domain, client));
+        IdentityProvider provider = identityProviderManager.getIdentityProvider(source);
         try {
-            Client client = context.get(ConstantKeys.CLIENT_CONTEXT_KEY);
-            User user = getUser(context, client);
-            String source =  Optional.ofNullable(user.getSource())
-                    .orElseGet(() -> UserRegistrationIdpResolver.getRegistrationIdp(domain, client));
-            IdentityProvider provider = identityProviderManager.getIdentityProvider(source);
-
             passwordService.validate(password, passwordPolicyManager.getPolicy(client, provider).orElse(null), user);
             context.next();
         } catch (InvalidPasswordException e) {
             Optional.ofNullable(context.request().getParam(Parameters.CLIENT_ID)).ifPresent(t -> queryParams.set(Parameters.CLIENT_ID, t));
+            if (user.getReferenceId() != null && user.getReferenceType() != null) {
```

Comment:
TODO: if the user doesn't have reference, you can add it using the domain instance and fix the UserAuditBuilder to not build the reference if the attributes are null (https://github.com/gravitee-io/gravitee-access-management/blob/master/gravitee-am-service/src/main/java/io/gravitee/am/service/reporter/builder/management/UserAuditBuilder.java#L64)

`auditService.report(AuditBuilder.builder(UserAuditBuilder.class).type(EventType.USER_PASSWORD_VALIDATION).user(user).reference(Reference.domain(domain.getId()).throwable(exception));`